### ISSUE TITLE
Kotlin Support

### ISSUE THREAD

**Issue by lenguyenthanh** [fabioCollini/DaggerMock#42]
Because most of classes and methods in Kotlin are `final` by default. So we need to mark all provide methods as `open` to use with DaggerMock. It's really a noise. 

Mockito 2 already support mock Kotlin class without `open` keyword. I wonder if we can do something similar with DaggerMock?

**Comment by plastiv** [comment]
Shouldn't just enabling MockMaker work? http://hadihariri.com/2016/10/04/Mocking-Kotlin-With-Mockito/
DaggerMock just delegates mock creation to mockito

**Comment by lenguyenthanh** [comment]
I enabled MockMaker already. It throws this exception:
```
java.lang.RuntimeException: The following methods must be non final:
public final android.app.Application com.lookata.app.AppModule.providesApplication()
public final android.content.res.Resources com.lookata.app.AppModule.providesResource()
public final com.lenguyenthanh.common.rx.RxSchedulers com.lookata.app.AppModule.providesRxSchedulers()

	at it.cosenonjaviste.daggermock.ErrorsFormatter.throwExceptionOnErrors(ErrorsFormatter.java:42)
	at it.cosenonjaviste.daggermock.ErrorsFormatter.throwExceptionOnErrors(ErrorsFormatter.java:37)
	at it.cosenonjaviste.daggermock.ModuleOverrider.checkMethodsVisibility(ModuleOverrider.java:86)
	at it.cosenonjaviste.daggermock.ModuleOverrider.override(ModuleOverrider.java:50)
	at it.cosenonjaviste.daggermock.DaggerMockRule.initComponent(DaggerMockRule.java:228)
	at it.cosenonjaviste.daggermock.DaggerMockRule.setupComponent(DaggerMockRule.java:124)
	at it.cosenonjaviste.daggermock.DaggerMockRule.access$000(DaggerMockRule.java:36)
	at it.cosenonjaviste.daggermock.DaggerMockRule$1.evaluate(DaggerMockRule.java:104)
	at org.robolectric.internal.SandboxTestRunner$2.evaluate(SandboxTestRunner.java:209)
	at org.robolectric.internal.SandboxTestRunner.runChild(SandboxTestRunner.java:109)
	at org.robolectric.internal.SandboxTestRunner.runChild(SandboxTestRunner.java:36)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.robolectric.internal.SandboxTestRunner$1.evaluate(SandboxTestRunner.java:63)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:117)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:42)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:262)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:84)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:147)
```

This is how I use DaggerMock
```
class MockServerRule : DaggerMockRule<App.AppComponent>(App.AppComponent::class.java, AppModule(RobolectricUnitTestRunner.app()), SupportModule()) {

    val mockWebServer = MockWebServer()

    init {

        try {
            mockWebServer.start()
        } catch (e: IOException) {
            e.printStackTrace()
        }

        val configuration = Mockito.mock<Configuration>(Configuration::class.java)
        whenever(configuration.getBaseApiUrl()).thenReturn(mockWebServer.url("").toString())

        provides<Configuration>(Configuration::class.java, configuration)
    }
}

@RunWith(RobolectricUnitTestRunner::class)
class ItemApiTest {
    @get:Rule
    val mockServerRule = MockServerRule().set {
        appComponent ->
        itemApi = appComponent.itemApi()
    } as MockServerRule

    private lateinit var itemApi: ItemApi
```

**Comment by lenguyenthanh** [comment]
DaggerMock will work excellently for Kotlin without this line: https://github.com/fabioCollini/DaggerMock/blob/master/lib/src/main/java/it/cosenonjaviste/daggermock/ModuleOverrider.java#L50

Or more specific, those lines:
```
if (isFinal(method.getModifiers())) {
      finalErrors.add(method.toString());
}
```

Reason is with MockMaker we can mock final stuffs. 

So this problem can be solved by solutions:
1. Remove it completely 😎 
2. Show warning for those final methods instead of throwing exception.

What do you think @fabioCollini ? If you want I'm happy to do a PR for second solutions.

**Comment by fabioCollini** [comment]
Hi, do you know if there is a way to detect whether MockMaker is used? This would be great so the error could be modified in a message like "remove final if you are using Java or use MockMaker if you are using Kotlin".
Thanks for your help!

**Comment by lenguyenthanh** [comment]
As far as I know, MockMaker can work with both Java and Kotlin. So the message should be like "Use Mocker if you are using final."

The problem is I don't know how to detect whether MockMaker is used. Maybe we need to look at Mockito source code. I'll try to investigate tomorrow.

**Comment by fabioCollini** [comment]
Maybe you can check if the file `org.mockito.plugins.MockMaker` exists but I am not sure this works in JUnit and Espresso tests. A simpler solution can be adding a new method `checkFinalMethods` on `DaggerMockRule` class and disable the check based on this value.
In this way if you use MockMaker you can just invoke this method.

**Comment by lenguyenthanh** [comment]
I just try final methods without MockMaker plugins. It throw exception:
```
org.mockito.exceptions.base.MockitoException: 
Cannot mock/spy class com.lookata.app.AppModule
Mockito cannot mock/spy because :
 - final class

	at it.cosenonjaviste.daggermock.ModuleOverrider.override(ModuleOverrider.java:64)
	at it.cosenonjaviste.daggermock.DaggerMockRule.initComponent(DaggerMockRule.java:228)
	at it.cosenonjaviste.daggermock.DaggerMockRule.setupComponent(DaggerMockRule.java:124)
	at it.cosenonjaviste.daggermock.DaggerMockRule.access$000(DaggerMockRule.java:36)
	at it.cosenonjaviste.daggermock.DaggerMockRule$1.evaluate(DaggerMockRule.java:104)
	at com.lenguyenthanh.common.MockServerRule$apply$1.evaluate(MockServerRule.kt:38)
	at org.robolectric.internal.SandboxTestRunner$2.evaluate(SandboxTestRunner.java:209)
	at org.robolectric.internal.SandboxTestRunner.runChild(SandboxTestRunner.java:109)
	at org.robolectric.internal.SandboxTestRunner.runChild(SandboxTestRunner.java:36)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.robolectric.internal.SandboxTestRunner$1.evaluate(SandboxTestRunner.java:63)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:117)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:42)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:262)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:84)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:147)
```
So I think the easiest way to solve this problem is:
1. Remove final methods check
2. Try catch this line https://github.com/fabioCollini/DaggerMock/blob/master/lib/src/main/java/it/cosenonjaviste/daggermock/ModuleOverrider.java#L64 with the exception above.

**Comment by lenguyenthanh** [comment]
Or we can use this method in Mockito: 
```
    public static TypeMockability typeMockabilityOf(Class<?> type) {
      return mockMaker.isTypeMockable(type);
    }
```

**Comment by lenguyenthanh** [comment]
I think this is even better: 
```
private <T> void checkMockability(T module) {
        MockMaker.TypeMockability typeMockability = MockUtil.typeMockabilityOf(module.getClass());
        if(typeMockability.mockable()) return;;
        Method[] methods = module.getClass().getDeclaredMethods();
        List<String> visibilityErrors = new ArrayList<>();
        List<String> finalErrors = new ArrayList<>();
        List<String> staticErrors = new ArrayList<>();
        for (Method method : methods) {
            if (method.isAnnotationPresent(Provides.class)) {
                if (!isPublic(method.getModifiers()) && !isProtected(method.getModifiers())) {
                    visibilityErrors.add(method.toString());
                }
                if (isStatic(method.getModifiers()) && overriddenObjectsMap.getProvider(method) != null) {
                    staticErrors.add(method.toString());
                }
            }
        }
        ErrorsFormatter.throwExceptionOnErrors("The following methods must be declared public or protected", visibilityErrors);
        ErrorsFormatter.throwExceptionOnErrors("The following methods must be non final", finalErrors);
        ErrorsFormatter.throwExceptionOnErrors("The following methods must be non static", staticErrors);
    }
```

**Comment by lenguyenthanh** [comment]
Just check it again
```
MockMaker.TypeMockability typeMockability = MockUtil.typeMockabilityOf(module.getClass());
        if(typeMockability.mockable()) return;
```
Code above doesn't work, `typeMockability.mockable()` always returns true. I'll take a look tomorrow.

**Comment by fabioCollini** [comment]
It seems to be a good solution, just a question: which is the mockito version that introduced MockMaker? DaggerMock works also with mockito 1, so if that class is not available on Mockito 1 the check should be executed using reflection

**Comment by lenguyenthanh** [comment]
`MockMaker` was introduced in version 1.9.5 which is quite long time a go (around 2012). But `ByteBuddyMockMaker` was introduced in version 2.0 only. 

So The new solution is:
```
   private Boolean isUsedMockMaker() {
        if(!isMockitoVersion2()) return false;
        MockMaker mockMaker = Plugins.getMockMaker();
        return !(mockMaker instanceof ByteBuddyMockMaker);
    }

    private Boolean isMockitoVersion2() {
        try {
            Class.forName("org.mockito.internal.creation.bytebuddy.ByteBuddyMockMaker");
            return true;
        } catch (ClassNotFoundException e) {
        }
        return false;
    }
```

Do you have any idea?

**Comment by lenguyenthanh** [comment]
@fabioCollini how do you think about the solution? I'm happy to do the PR.

**Comment by fabioCollini** [comment]
It seems good, if you can create the PR it will be great! So I can try it in a couple of project before creating a new release.
Thanks again for your help!

**Comment by lenguyenthanh** [comment]
OK, I will do it.

**Comment by Douvi** [comment]
Hey everyone, 

After using the lenguyenthanh 's PR: 
`androidTestCompile 'com.github.lenguyenthanh:DaggerMock:50b0dc76d5689820d67964efed050f2f7fc197c9'`

Here my Module
```
@Module(includes = arrayOf(HouseModule::class))
class PeopleMock {

	@Provides
	@ApplicationScope
	fun providePeople(impl: PeopleMock):  People = impl
}
```

How I call it from my test:
```
@Rule @JvmField var daggerRule = DaggerMockRule(GlobalComponent::class.java, PeopleMock()).set({ globalComponent ->
//my Code
	})
```

I got an exception:
```
org.mockito.exceptions.base.MockitoException:
Mockito cannot mock this class: class com.test.mock.PeopleMock.

Mockito can only mock non-private & non-final classes.
If you're not sure why you're getting this error, please report to the mailing list.



IMPORTANT INFORMATION FOR ANDROID USERS:

The regular Byte Buddy mock makers cannot generate code on an Android VM!
To resolve this, please use the 'mockito-android' dependency for your application:
http://search.maven.org/#search%7Cga%7C1%7Ca%3A%22mockito-android%22%20g%3A%22org.mockito%22

Java               : 0.9
JVM vendor name    : The Android Project
JVM vendor version : 2.1.0
JVM name           : Dalvik
JVM version        : 0.9
JVM info           : null
OS name            : Linux
OS version         : 3.10.0+


Underlying exception : java.lang.UnsupportedOperationException: Cannot define class using reflection
at it.cosenonjaviste.daggermock.ModuleOverrider.override(ModuleOverrider.java:69)
at it.cosenonjaviste.daggermock.DaggerMockRule.initComponent(DaggerMockRule.java:228)
at it.cosenonjaviste.daggermock.DaggerMockRule.setupComponent(DaggerMockRule.java:124)
at it.cosenonjaviste.daggermock.DaggerMockRule.access$000(DaggerMockRule.java:36)
at it.cosenonjaviste.daggermock.DaggerMockRule$1.evaluate(DaggerMockRule.java:104)
at android.support.test.internal.statement.UiThreadStatement.evaluate(UiThreadStatement.java:55)
at android.support.test.rule.ActivityTestRule$ActivityStatement.evaluate(ActivityTestRule.java:270)
at br.com.concretesolutions.requestmatcher.RequestMatcherRule$1.evaluate(RequestMatcherRule.java:305)
at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
at org.junit.rules.RunRules.evaluate(RunRules.java:20)
at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
at org.junit.runners.Suite.runChild(Suite.java:128)
at org.junit.runners.Suite.runChild(Suite.java:27)
at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
at android.support.test.internal.runner.TestExecutor.execute(TestExecutor.java:59)
at android.support.test.runner.AndroidJUnitRunner.onStart(AndroidJUnitRunner.java:262)
at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:1932)
Caused by: java.lang.UnsupportedOperationException: Cannot define class using reflection
at net.bytebuddy.dynamic.loading.ClassInjector$UsingReflection$Dispatcher$Unavailable.defineClass(ClassInjector.java:821)
at net.bytebuddy.dynamic.loading.ClassInjector$UsingReflection.inject(ClassInjector.java:185)
at net.bytebuddy.dynamic.loading.ClassLoadingStrategy$Default$InjectionDispatcher.load(ClassLoadingStrategy.java:187)
at net.bytebuddy.dynamic.TypeResolutionStrategy$Passive.initialize(TypeResolutionStrategy.java:79)
at net.bytebuddy.dynamic.DynamicType$Default$Unloaded.load(DynamicType.java:4376)
at org.mockito.internal.creation.bytebuddy.SubclassBytecodeGenerator.mockClass(SubclassBytecodeGenerator.java:94)
at org.mockito.internal.creation.bytebuddy.TypeCachingBytecodeGenerator$1.call(TypeCachingBytecodeGenerator.java:37)
at org.mockito.internal.creation.bytebuddy.TypeCachingBytecodeGenerator$1.call(TypeCachingBytecodeGenerator.java:34)
at net.bytebuddy.TypeCache.findOrInsert(TypeCache.java:138)
at net.bytebuddy.TypeCache$WithInlineExpunction.findOrInsert(TypeCache.java:346)
at net.bytebuddy.TypeCache.findOrInsert(TypeCache.java:161)
at net.bytebuddy.TypeCache$WithInlineExpunction.findOrInsert(TypeCache.java:355)
at org.mockito.internal.creation.bytebuddy.TypeCachingBytecodeGenerator.mockClass(TypeCachingBytecodeGenerator.java:32)
at org.mockito.internal.creation.bytebuddy.SubclassByteBuddyMockMaker.createMockType(SubclassByteBuddyMockMaker.java:71)
at org.mockito.internal.creation.bytebuddy.SubclassByteBuddyMockMaker.createMock(SubclassByteBuddyMockMaker.java:42)
at org.mockito.internal.creation.bytebuddy.ByteBuddyMockMaker.createMock(ByteBuddyMockMaker.java:26)
at org.mockito.internal.util.MockUtil.createMock(MockUtil.java:35)
at org.mockito.internal.MockitoCore.mock(MockitoCore.java:65)
at org.mockito.Mockito.mock(Mockito.java:1729)
at org.mockito.Mockito.mock(Mockito.java:1702)
... 32 more
Caused by: java.lang.IllegalStateException: This JVM's version string does not seem to be valid: 0
at net.bytebuddy.ClassFileVersion$VersionLocator$ForLegacyVm.locate(ClassFileVersion.java:368)
at net.bytebuddy.ClassFileVersion.ofThisVm(ClassFileVersion.java:147)
at net.bytebuddy.dynamic.loading.ClassInjector$UsingReflection$Dispatcher$CreationAction.run(ClassInjector.java:301)
at net.bytebuddy.dynamic.loading.ClassInjector$UsingReflection$Dispatcher$CreationAction.run(ClassInjector.java:290)
at java.security.AccessController.doPrivileged(AccessController.java:41)
at net.bytebuddy.dynamic.loading.ClassInjector$UsingReflection.<clinit>(ClassInjector.java:70)
at net.bytebuddy.dynamic.loading.ClassLoadingStrategy$Default$InjectionDispatcher.load(ClassLoadingStrategy.java:184)
... 49 more
```

Can you tell me what I'm missing ?

Thanks

**Comment by lenguyenthanh** [comment]
Are you enable Mockito Mockmaker Inline plugin? It seems using `ByteBuddyMockMaker` not `InlineByteBuddyMockMaker`. Pls see: http://hadihariri.com/2016/10/04/Mocking-Kotlin-With-Mockito/.

But DaggerMock should handle this case and give a better error message any way.

Do you have sample code for this problem? I'll take a look at it. Thanks

**Comment by fabioCollini** [comment]
I created a new [branch](https://github.com/fabioCollini/DaggerMock/tree/kotlin) with a kotlin example and I get a similar problem running Android tests (JVM tests work). Have you ever used MockMaker for Espresso tests? Can you take a look at that branch?

**Comment by lenguyenthanh** [comment]
Ok, I'll take a look at it. I haven't tried MockMaker for Espresso tests yet, but will use for sure.

**Comment by lenguyenthanh** [comment]
I have created a PR for your branch. It should work now.

The problem was `inline-mock-maker` doesn't work on Android (https://static.javadoc.io/org.mockito/mockito-core/2.8.9/org/mockito/Mockito.html#mockito). So that we need to use `mockito-android` dependency and DexOpener library(https://github.com/tmurakami/dexopener) to mock final classes, methods.

I think the solution here, is better document and example :).

**Comment by danielocampo2** [comment]
Hey, that's a really cool addition, will try it out soon.

As a side note, there is an annotation available that you can add to your classes: @AllOpen. It will make all the methods and class open. We use it for our Dagger modules.

**Comment by fabioCollini** [comment]
@danielocampo2 the annotation would be great but I can't find it. Is it a standard a standard annotation or is it available with a library? Do you have a reference to the documentation?

**Comment by PavelSynek** [comment]
You have to add the compiler plugin and declare your own annotation
https://kotlinlang.org/docs/reference/compiler-plugins.html#all-open-compiler-plugin

**Comment by danielocampo2** [comment]
Yes, forgot that important detail! Thanks @PavelSynek for clarifying.

**Comment by fabioCollini** [comment]
Ok, but do you need to annotate all the modules, the components and the classes you want to mock? Is there an overhead on production code?
However I'll try it to check whether DaggerMock works with this configuration.
Thanks for the link!

**Comment by danielocampo2** [comment]
@fabioCollini we only use it for DaggerMock purposes so the only classes that we annotate are the Dagger modules that DaggerMock needs to override.
The rest of our code makes use of interfaces we usually don't need it.

**Comment by PaulWoitaschek** [comment]
You can use the annotation to annotate annotations.

```kotlin
@Target(AnnotationTarget.ANNOTATION_CLASS)
annotation class AllOpen
```

Then you can use another annotation you use to mark your classes and simply not apply the AllOpen to that class in release builds.

So your `debug/java/src/main...`will be
```kotlin
@AllOpen
@Target(AnnotationTarget.CLASS)
annotation class OpenForTesting
```

And your `release/java/src/main...`will be

```kotlin
@Target(AnnotationTarget.CLASS)
annotation class OpenForTesting
```

That way in release builds everything stays final, if that is the concern.

**Comment by fabioCollini** [comment]
I have finally released a version 0.7.0 with all the modification discussed in this issue. I have created a new demo project using All Open Kotlin plugin, it works but I annotated a lot of classes.
Thanks to everyone for your help!

**Comment by lenguyenthanh** [comment]
Great, Thanks @fabioCollini.
